src/pybind/mgr/cephadm/services/monitoring.py

```diff
@@ -256,86 +255,46 @@ def generate_config(
-            # disabled in order to be consistent with _calc_daemon_deps().
-            deps.append(dd.name())
-            if not port:
-                continue
-            if dd.daemon_id == self.mgr.get_mgr_id():
-                continue
-            assert dd.hostname is not None
-            addr = self._inventory_get_fqdn(dd.hostname)
-            mgr_scrape_list.append(build_url(host=addr, port=port).lstrip('/'))
-
-        # scrape node exporters
-        nodes = []
-        for dd in self.mgr.cache.get_daemons_by_service('node-exporter'):
-            assert dd.hostname is not None
-            deps.append(dd.name())
-            addr = dd.ip if dd.ip else self._inventory_get_fqdn(dd.hostname)
-            port = dd.ports[0] if dd.ports else 9100
-            nodes.append({
-                'hostname': dd.hostname,
-                'url': build_url(host=addr, port=port).lstrip('/')
-            })
+                srv_end_point = f'https://{p_result.hostname}:{sd_port}/sd/prometheus/sd-config?'
 
-        # scrape alert managers
-        alertmgr_targets = []
-        for dd in self.mgr.cache.get_daemons_by_service('alertmanager'):
-            assert dd.hostname is not None
-            deps.append(dd.name())
-            addr = dd.ip if dd.ip else self._inventory_get_fqdn(dd.hostname)
-            port = dd.ports[0] if dd.ports else 9093
-            alertmgr_targets.append("'{}'".format(build_url(host=addr, port=port).lstrip('/')))
-
-        # scrape haproxies
-        haproxy_targets = []
-        for dd in self.mgr.cache.get_daemons_by_type('ingress'):
-            if dd.service_name() in self.mgr.spec_store:
-                spec = cast(IngressSpec, self.mgr.spec_store[dd.service_name()].spec)
-                assert dd.hostname is not None
-                deps.append(dd.name())
-                if dd.daemon_type == 'haproxy':
-                    addr = self._inventory_get_fqdn(dd.hostname)
-                    haproxy_targets.append({
-                        "url": f"'{build_url(host=addr, port=spec.monitor_port).lstrip('/')}'",
-                        "service": dd.service_name(),
-                    })
+        mgr_prometheus_sd_url = f'{srv_end_point}service=mgr-prometheus'
+        node_exporter_sd_url = f'{srv_end_point}service=node-exporter'
+        alertmanager_sd_url = f'{srv_end_point}service=alertmanager'
+        haproxy_sd_url = f'{srv_end_point}service=haproxy'
 
         # generate the prometheus configuration
         context = {
-            'alertmgr_targets': alertmgr_targets,
-            'mgr_scrape_list': mgr_scrape_list,
-            'haproxy_targets': haproxy_targets,
-            'nodes': nodes,
+            'mgr_prometheus_sd_url': mgr_prometheus_sd_url,
+            'node_exporter_sd_url': node_exporter_sd_url,
+            'alertmanager_sd_url': alertmanager_sd_url,
+            'haproxy_sd_url': haproxy_sd_url,
         }
+
```

Comment:
Yeah, I'm aware of it. Before this change the configuration was static and each time a new service was deployed we had to generate the Prometheus yaml configuration file to include the new section. This has changed and now our configuration is obtained by using SD. Prometheus as I've seen on my testing handles the empty targets correctly and basically don't show them on the UI.

IMHO the configuration is not confusing. Basically it lists all the supported services and the URLs to discover if this service has been configured or not. When the service is not configured an empty list is returned. When the service is configured valid targets are returned. As I said before Prometheus seems to handle this correctly and doesn't show entries who have no targets so having these sections in our config file is not harmful in my opinion.

Reply:
Different perspectives :) I try and think about it from a less experienced admin's perspective. Looking at a config that declares a scrape job for a non-existing service make make them question whether the job is a remnant from an old deployment, or a failed deployment - that's why I would advocate to only show a config that represents the *current* state. Also, by simply showing everything you're reliant upon the default behavior of prometheus - this may never change, but it's not anything we can control!

Reply:
Okay, you convinced me .. specially with the argument of depending on the Prometheus behavior :)

I changed the code in the last commit to only includes the sections corresponding to configured daemons/services.